[zrei] iP

zrei opened the pull request:
# PRESENTING THE ILLUSTRIOUS DUKE

> "I can't keep track of everything that I have to do!" - Every CS student, ever

If you've ever found yourself at a loss on how to keep track of your **humongous, ever-increasing mountain of tasks**, then Duke is *perfect* for you!

Duke can do many ~god-like, reality-shattering~ amazing things, including but not limited to:
- Store your tasks (in three different categories! ***Incredible, right?*** :innocent:)
- Mark your tasks as done or not done!
- ~Control reality~ Remember your tasks in-between sessions!

You, too, can get access to this awesomeness through a few simple steps.
1. Download it from [here](https://github.com/zrei/ip/releases/tag/A-Jar)
2. ~Give up your soul~ Start it up!

Start Duke with this simple `command`:
```
java -jar duke.jar
```

So, what are you waiting for? 
- [ ] Become a Duke user! You definitely won't regret it!

---

And here's some random code that I have to include for some arbitrary reason. It's the classic `helloWorld` function!
```java
public static void helloWorld() {
    System.out.println("Hello World");
}
```

~Can you believe that this prints `Hello World`?~

Commit:
Add Gradle support

zrei committed:
Implement level-1

Make duke echo commands and exit when user types "bye".

zrei committed:
Implement Level-2

Duke can now store text and display the stored text to users.

zrei committed:
Add Task class

zrei committed:
Use tasks in Duke

zrei committed:
Implement Level-3 of Duke

Tasks can now be marked as done or unmarked as done.

zrei committed:
Add Deadline.java

Implement Deadline class that extends from Task.

zrei committed:
Add Todo.java

Implement Todo class that extends from Task.

zrei committed:
Add Event.java

Implement Event class that extends from Task and make Task an abstract
class.

zrei committed:
Implement adding of todos to task list

zrei committed:
Implement adding of deadlines to task list

zrei committed:
Implement adding of events to task list

zrei committed:
Add additional line to list command response

zrei committed:
Create test case for automated text UI testing

zrei committed:
Change permissions on text-ui-test/runtest.sh

zrei committed:
Create DukeException.java

Implement exceptions.

zrei committed:
Handle certain exceptions

zrei committed:
Implement deleting of tasks from the task list

zrei committed:
Implement catching of more errors

zrei committed:
Separate processing of commands into different functions

zrei committed:
Add function to convert task to a string representation for storage

zrei committed:
Implement reading from and writing to storage

zrei committed:
Merge branch 'branch-Level-7'

zrei committed:
Add date support to Deadlines and Events

zrei committed:
Merge branch 'branch-Level-8'

zrei committed:
Create directory data only when directory does not exist

Previously, Files.createDirectory(dirPath) was throwing an
IOException when the directory data already existed.

Creating the directory data only when it does not exist
prevents the IOException from being thrown. The saved task
list is now loaded correctly.

zrei committed:
Merge branch 'add-gradle-support'

zrei committed:
Update .gitignore

Add VS code and local storage files to .gitignore.

zrei committed:
Add checkstyle support to gradle

zrei committed:
Add backup of old Duke.java

zrei committed:
Add Storage.java

zrei committed:
Add method to convert task to a string to be saved in storage

zrei committed:
Add TaskList.java

zrei committed:
Add Ui.java

zrei committed:
Add abstract class Command.java

zrei committed:
Add individual command classes

zrei committed:
Add Parser.java

zrei committed:
Refactor Duke

zrei committed:
Separate duke into packages

Task class and its subclasses go into the task package.
Exceptions go into the exception package.
Command class and its subclasses go into the command package.
All packages and all other files are in the duke package.

zrei committed:
Update .gitignore to ignore all .class files

zrei committed:
Shift task package into duke

zrei committed:
Add JUnit tests

Add JUnit tests for Todo and Deadline classes.

zrei committed:
Update mainClassName for shadowJar

zrei committed:
Add JavaDoc comments

zrei committed:
Update code to meet coding standard

zrei committed:
Add method to check if task's name contains a keyword

zrei committed:
Add function to filter Task List by keyword

zrei committed:
Add find command

zrei committed:
Add support for parsing find command

zrei committed:
Merge branch 'branch-A-JavaDoc'

zrei committed:
Merge branch 'branch-A-CodingStandard'

zrei committed:
Merge branch 'branch-Level-9'

zrei committed:
Display different response when there are no tasks in the list

zrei committed:
Update JavaDoc to match Java coding standard

zrei committed:
Update code style

zrei committed:
Set up JavaFX

zrei committed:
Add image files

zrei committed:
Add GUI for Duke

zrei committed:
Separate GUI into separate class Gui

zrei committed:
Create package ui

zrei committed:
Rename ui package to gui

zrei committed:
Use FXML in JavaFX

zrei committed:
Add padding to text

zrei committed:
Merge branch 'branch-Level-10'

zrei committed:
Update GUI appearance and text responses

zrei committed:
Add command and task type enums

zrei committed:
Use varargs for creation of tasks

zrei committed:
Update GUI

zrei committed:
Update enum function

zrei committed:
Round image corners

zrei committed:
Update variable names

zrei committed:
Add help command

zrei committed:
Allow find to match multiple keywords

zrei committed:
Show goodbye message upon entering bye

potty10 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,103 @@
+            return new Deadline(name, by);
+        case EVENT:
+            String from = arguments[1];
+            String to = arguments[2];
```
Perhaps using nouns for this variables name, such as `startTime` is better?

potty10 commented on src/main/java/duke/gui/DialogBox.java:
```diff
@@ -0,0 +1,115 @@
+*/
+public class DialogBox extends HBox {
+
+    private static final float WRAPPING_WIDTH = 250;
```
I like the usage and naming of constants here! Understanding it is very convenient.

potty10 commented on src/test/java/duke/task/DeadlineTest.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @Test
+    public void deadlineCreation_invalidDate_exceptionThrown() {
```
Nice naming of test cases here!

potty10 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,30 @@
+    private LocalDate from;
+    private LocalDate to;
+
+    public Event(String name, String from, String to) throws DateTimeParseException {
```
Should header comments be added here to describe the parameters of the constructor? I feel that it will be useful.

potty10 commented on src/main/java/duke/command/CommandType.java:
```diff
@@ -0,0 +1,61 @@
+    BYE, MARK, UNMARK, DEADLINE, TODO, EVENT, LIST, FIND, DELETE, HELP;
+
+    public static String getAllCommandFormats() {
+        String allCommandFormats = "";
```
Perhaps a better variable name is `allCommandFormatString`, since plural form is usually for collection of objects?

zrei replied:
Thanks for the suggestion! I think that's a good idea.

potty10 reviewed (commented):
LGTM. Just some nits to fix.

zrei committed:
Add assertions

Classes lack assertions.

This causes a lack of documentation of important assumptions that
should hold at various stages of runtime.

Adding assertions allows runtime bugs to be identified more easily.

Let's add assertions to multiple classes.

zrei committed:
Change Event class variable names

Event class has variables from and to.

These variables names are not clear in their meaning.

Let's change these variables names to startDate and endDate
to make their meaning clearer.

zrei committed:
Change method name for getting all commands

zrei committed:
Improve quality

zrei committed:
Add additional header comments

zrei committed:
Merge pull request #1 from zrei/branch-A-Assertions

Add assertions

zrei committed:
Merge branch 'master' into branch-A-CodeQuality

zrei committed:
Merge pull request #2 from zrei/branch-A-CodeQuality

Improve code quality

zrei committed:
Set up CI

zrei committed:
Help command: launch a separate window

When a help command is issued, Duke outputs a long response
containing all commands it recognises.

This is hard to read due to the text being longer than the
window length, and needing the user to scroll up and down to
find the command they want to execute.

Let's add a separate window to contain all the commands that Duke
recognises, and launch this window when a help command is issued.

TODO:
* Improve the GUI of the help window

zrei committed:
Merge pull request #3 from zrei/branch-BCD-Extension

Help command: launch a separate window

AngJunYang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,70 @@
+            fw.write(taskList);
+            fw.close();
+        } catch (IOException e) {
+            throw new DukeException("I couldn't write to storage. I tripped past the entranceway.");
```
I like these personalized exception messages!

zrei committed:
Shift resources into main folder to work with Gradle

AngJunYang commented on src/main/java/duke/command/CommandType.java:
```diff
@@ -0,0 +1,67 @@
+package duke.command;
+
+/** Enum to represent commands */
+public enum CommandType {
```
I like the use of enums in your implementation of Commands!

AngJunYang commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,28 @@
+import duke.gui.SpriteEmotion;
+import duke.task.Task;
+
+public class DeleteCommand extends Command {
```
Perhaps some descriptive header commends for this class?

AngJunYang commented on src/main/java/duke/gui/DialogBox.java:
```diff
@@ -0,0 +1,115 @@
+    private static final float ROUNDED_CORNERS_LENGTH = 25;    
+
+    private static final Color DUKE_COLOR = Color.CORNFLOWERBLUE;
+    private static final Color USER_COLOR = Color.CRIMSON;
```
Perhaps associated constants should have a constant prefix?

zrei replied:
Thanks for pointing this out! I didn't notice this before now.

AngJunYang reviewed (commented):
I like the use of GIFs and the personalized feel of the project! Looks good with few coding issues, keep up the good work!

zrei committed:
Change chatbot name

zrei committed:
Update JavaDoc

zrei committed:
Add user guide

zrei committed:
Update user guide

zrei committed:
Add theme to github page

zrei committed:
Update user guide

zrei committed:
Add link to help window

zrei committed:
Add images to readme

zrei committed:
Add more images to user guide

zrei committed:
Align image

zrei committed:
Align images

zrei committed:
Update user guide

zrei committed:
Style images

zrei committed:
Adjust height of images

zrei committed:
Adjust heights

zrei committed:
Adjust heights

zrei committed:
Add scroll bar to help window

zrei committed:
Update README.md
